File: libmach1spatial/api_encode/src/Mach1EncodeEmscripten.cpp
#include <emscripten/bind.h>

#ifdef M1ENCODE_INLINE_DECODE
#    include "Mach1DecodeCAPI.h"
#endif

#include "../include/Mach1Encode.h"
#include "Mach1EncodeCAPI.h"

using namespace emscripten;

EMSCRIPTEN_BINDINGS(Mach1Encode) {
    enum_<Mach1EncodeInputMode>("Mach1EncodeInputMode")
        .value("Mono", Mono)
        .value("Stereo", Stereo)
        .value("LCR", LCR)
        .value("Quad", Quad)
        .value("LCRS", LCRS)
        .value("AFormat", AFormat)
        .value("FiveDotZero", FiveDotZero)
        .value("FiveDotOneFilm", FiveDotOneFilm)
        .value("FiveDotOneDTS", FiveDotOneDTS)
        .value("FiveDotOneSMTPE", FiveDotOneSMTPE)
        .value("BFOAACN", BFOAACN)
        .value("BFOAFUMA", BFOAFUMA)
        .value("B2OAACN", B2OAACN)
        .value("B2OAFUMA", B2OAFUMA)
        .value("B3OAACN", B3OAACN)
        .value("B3OAFUMA", B3OAFUMA);

    enum_<Mach1EncodeOutputMode>("Mach1EncodeOutputMode")
        .value("M1Spatial_4", M1Spatial_4)
        .value("M1Spatial_8", M1Spatial_8)
        .value("M1Spatial_14", M1Spatial_14);

    enum_<Mach1EncodePannerMode>("Mach1EncodePannerMode")
        .value("IsotropicLinear", IsotropicLinear)
        .value("IsotropicEqualPower", IsotropicEqualPower)
        .value("PeriphonicLinear", PeriphonicLinear);

#ifdef M1ENCODE_INLINE_DECODE
    enum_<Mach1DecodeMode>("Mach1DecodeMode")
        .value("M1Spatial_4", M1Spatial_4)
        .value("M1Spatial_8", M1Spatial_8)
        .value("M1Spatial_14", M1Spatial_14);
#endif

    register_vector<float>("VectorFloat");
    register_vector<std::string>("VectorString");
    register_vector<Mach1Point3D>("VectorMach1Point3D");
    register_vector<std::vector<float> >("VectorVectorFloat");

    value_object<Mach1Point3D>("Mach1Point3D")
        .field("x", &Mach1Point3D::x)
        .field("y", &Mach1Point3D::y)
        .field("z", &Mach1Point3D::z);

    class_<Mach1Encode<float> >("Mach1EncodeInternal")
        .constructor<>()
        .function("getPoints", &Mach1Encode<float>::getPoints)
        .function("getGains", &Mach1Encode<float>::getGains)
        .function("getPointsNames", &Mach1Encode<float>::getPointsNames)
        .function("getGainsForInputChannelNamed", &Mach1Encode<float>::getGainsForInputChannelNamed)

        .function("generatePointResults", &Mach1Encode<float>::generatePointResults)
        .function("getPointsCount", &Mach1Encode<float>::getPointsCount)

#ifdef M1ENCODE_INLINE_DECODE
        .function("getResultingCoeffsDecoded", &Mach1Encode<float>::getResultingCoeffsDecoded)
#endif
        .function("getInputMode", &Mach1Encode<float>::getInputMode)
        .function("getOutputMode", &Mach1Encode<float>::getOutputMode)
        .function("getInputChannelsCount", &Mach1Encode<float>::setInputMode)
        .function("getOutputChannelsCount", &Mach1Encode<float>::setOutputMode)

        .function("setInputMode", &Mach1Encode<float>::setInputMode)
        .function("setOutputMode", &Mach1Encode<float>::setOutputMode)

        .function("setAzimuth", &Mach1Encode<float>::setAzimuth)
        .function("setAzimuthDegrees", &Mach1Encode<float>::setAzimuthDegrees)
        .function("setAzimuthRadians", &Mach1Encode<float>::setAzimuthRadians)

        .function("setDiverge", &Mach1Encode<float>::setDiverge)

        .function("setElevation", &Mach1Encode<float>::setElevation)
        .function("setElevationDegrees", &Mach1Encode<float>::setElevationDegrees)
        .function("setElevationRadians", &Mach1Encode<float>::setElevationRadians)

        .function("setPannerMode", &Mach1Encode<float>::setPannerMode)
        .function("setFrontSurroundPerspective", &Mach1Encode<float>::setFrontSurroundPerspective)
        .function("setOutputGain", &Mach1Encode<float>::setOutputGain)
        .function("getOutputGain", &Mach1Encode<float>::getOutputGain)
        .function("getGainCompensationActive", &Mach1Encode<float>::getGainCompensationActive)
        .function("setGainCompensationActive", &Mach1Encode<float>::setGainCompensationActive)
        .function("getGainCompensation", &Mach1Encode<float>::getGainCompensation)

        .function("setAutoOrbit", &Mach1Encode<float>::setAutoOrbit)
        .function("setOrbitRotation", &Mach1Encode<float>::setOrbitRotation)
        .function("setOrbitRotationDegrees", &Mach1Encode<float>::setOrbitRotationDegrees)
        .function("setOrbitRotationRadians", &Mach1Encode<float>::setOrbitRotationRadians)
        .function("setStereoSpread", &Mach1Encode<float>::setStereoSpread);
}

File: libmach1spatial/api_encode/src/Mach1EncodeEmscripten.js
Module['onRuntimeInitialized'] = function() {
    console.log('loaded Mach1Encode');

    Module['Mach1Encode'] = function() {
        this.Mach1EncodeInputMode = {
            "Mono": (Module).Mach1EncodeInputMode.Mono,
            "Stereo": (Module).Mach1EncodeInputMode.Stereo,
            "LCR": (Module).Mach1EncodeInputMode.LCR,
            "Quad": (Module).Mach1EncodeInputMode.Quad,
            "LCRS": (Module).Mach1EncodeInputMode.LCRS,
            "AFormat": (Module).Mach1EncodeInputMode.AFormat,
            "FiveDotZero": (Module).Mach1EncodeInputMode.otZero,
            "FiveDotOneFilm": (Module).Mach1EncodeInputMode.FiveDotOneFilm,
            "FiveDotOneDTS": (Module).Mach1EncodeInputMode.FiveDotOneDTS,
            "FiveDotOneSMTPE": (Module).Mach1EncodeInputMode.FiveDotOneSMTPE,
            "BFOAACN": (Module).Mach1EncodeInputMode.BFOAACN,
            "BFOAFUMA": (Module).Mach1EncodeInputMode.BFOAFUMA,
            "B2OAACN": (Module).Mach1EncodeInputMode.B2OAACN,
            "B2OAFUMA": (Module).Mach1EncodeInputMode.B2OAFUMA,
            "B3OAACN": (Module).Mach1EncodeInputMode.B3OAACN,
            "B3OAFUMA": (Module).Mach1EncodeInputMode.B3OAFUMA,
        };

        this.Mach1EncodeOutputMode = {
            "M1Spatial_4": (Module).Mach1EncodeOutputMode.M1Spatial_4,
            "M1Spatial_8": (Module).Mach1EncodeOutputMode.M1Spatial_8,
            "M1Spatial_12": (Module).Mach1EncodeOutputMode.M1Spatial_12,
            "M1Spatial_14": (Module).Mach1EncodeOutputMode.M1Spatial_14,
        };

        this.Mach1EncodePannerMode = {
            "IsotropicLinear": (Module).Mach1EncodePannerMode.IsotropicLinear,
            "IsotropicEqualPower": (Module).Mach1EncodePannerMode.IsotropicEqualPower,
            "PeriphonicLinear": (Module).Mach1EncodePannerMode.PeriphonicLinear,
        };

        this.Mach1DecodeMode = {
            "M1Spatial_4": (Module).Mach1DecodeMode.M1Spatial_4,
            "M1Spatial_8": (Module).Mach1DecodeMode.M1Spatial_8,
            "M1Spatial_12": (Module).Mach1DecodeMode.M1Spatial_12,
            "M1Spatial_14": (Module).Mach1DecodeMode.M1Spatial_14,
        };

        this.delete = function() {
            if (this._m1obj) {
                this._m1obj.delete();
                this._m1obj = null;
            }
        };

        this.getPoints = function() {
            if (this._m1obj) {
                var arr = this._m1obj.getPoints();
                var points = [];
                for (var i = 0; i < arr.size(); i++) {
                    points[i] = {
                        x: arr.get(i).x,
                        y: arr.get(i).y,
                        z: arr.get(i).z
                    };
                }
                arr.delete();
                arr = null;

                return points
            }
        };

        this.getGains = function() {
            if (this._m1obj) {
                var arr = this._m1obj.getGains();
                var gains = [];
                for (var i = 0; i < arr.size(); i++) {
                    gains[i] = [];
                    var _arr = arr.get(i);
                    for (var j = 0; j < _arr.size(); j++) {
                        gains[i][j] = _arr.get(j);
                    }
                    _arr.delete();
                    _arr = null;
                }
                arr.delete();
                arr = null;

                return gains;
            }
        };

        this.getResultingCoeffsDecoded  = function(decodeMode, decodeResult) {
            if (this._m1obj) {
                var vec = new(Module).VectorFloat();
                for (var i = 0; i < decodeResult.length; i++) {
                    vec.push_back(decodeResult[i]);
                }

                var arr = this._m1obj.getResultingCoeffsDecoded(decodeMode, vec);
                var volumes = [];
                for (var i = 0; i < arr.size(); i++) {
                    volumes[i] = arr.get(i);
                }
                arr.delete();
                arr = null;

                vec.delete();
                vec = null;

                return volumes;
            }
        };

        this.getPointsNames = function() {
            if (this._m1obj) {
                var arr = this._m1obj.getPointsNames();
                var pointsNames = [];
                for (var i = 0; i < arr.size(); i++) {
                    pointsNames[i] = arr.get(i);
                }
                arr.delete();
                arr = null;

                return pointsNames;
            }
        };

        this.getGainsForInputChannelNamed = function(channelName) {
            if (this._m1obj) {
                return this._m1obj.getGainsForInputChannelNamed(channelName);
            }
        };

        this.generatePointResults = function() {
            if (this._m1obj) {
                this._m1obj.generatePointResults();
            }
        };

        this.getPointsCount = function() {
            if (this._m1obj) {
                return this._m1obj.getPointsCount();
            }
        };

        this.getInputMode = function() {
            if (this._m1obj) {
                return this._m1obj.getInputMode();
            }
        };

        this.getOutputMode = function() {
            if (this._m1obj) {
                return this._m1obj.getOutputMode();
            }
        };

        this.getInputChannelsCount = function() {
            if (this._m1obj) {
                return this._m1obj.getInputChannelsCount();
            }
        };

        this.getOutputChannelsCount = function() {
            if (this._m1obj) {
                return this._m1obj.getOutputChannelsCount();
            }
        };

        this.setInputMode = function(inputMode) {
            if (this._m1obj) {
                this._m1obj.setInputMode(inputMode);
            }
        };

        this.setOutputMode = function(outputMode) {
            if (this._m1obj) {
                this._m1obj.setOutputMode(outputMode);
            }
        };

        this.setAzimuth = function(azimuth) {
            if (this._m1obj) {
                this._m1obj.setAzimuth(azimuth);
            }
        };

        this.setAzimuthDegrees = function(azimuth) {
            if (this._m1obj) {
                this._m1obj.setAzimuthDegrees(azimuth);
            }
        };

        this.setAzimuthRadians = function(azimuth) {
            if (this._m1obj) {
                this._m1obj.setAzimuthRadians(azimuth);
            }
        };

        this.setDiverge = function(diverge) {
            if (this._m1obj) {
                this._m1obj.setDiverge(diverge);
            }
        };

        this.setPitch = function(pitch) {
            if (this._m1obj) {
                this._m1obj.setElevationDegrees(pitch);
            }
        };

        this.setElevation = function(elevation) {
            if (this._m1obj) {
                this._m1obj.setElevation(elevation);
            }
        };

        this.setElevationDegrees = function(elevation) {
            if (this._m1obj) {
                this._m1obj.setElevationDegrees(elevation);
            }
        };

        this.setElevationRadians = function(elevation) {
            if (this._m1obj) {
                this._m1obj.setElevationRadians(elevation);
            }
        };

        this.setAutoOrbit = function(autoOrbit) {
            if (this._m1obj) {
                this._m1obj.setAutoOrbit(autoOrbit);
            }
        };

        this.setOrbitRotation = function(orbitRotation) {
            if (this._m1obj) {
                this._m1obj.setOrbitRotation(orbitRotation);
            }
        };

        this.setOrbitRotationDegrees = function(orbitRotation) {
            if (this._m1obj) {
                this._m1obj.setOrbitRotationDegrees(orbitRotation);
            }
        };

        this.setOrbitRotationRadians = function(orbitRotation) {
            if (this._m1obj) {
                this._m1obj.setOrbitRotationRadians(orbitRotation);
            }
        };

        this.setStereoSpread = function(stereoSpread) {
            if (this._m1obj) {
                this._m1obj.setStereoSpread(stereoSpread);
            }
        };

        this.setPannerMode = function(pannerMode) {
            if (this._m1obj) {
                this._m1obj.setPannerMode(pannerMode);
            }
        };

        this.setFrontSurroundPerspective = function(frontSurroundPerspective) {
            if (this._m1obj) {
                this._m1obj.setFrontSurroundPerspective(frontSurroundPerspective);
            }
        };

        this.setOutputGain = function(outputGainMultipler, isDecibel) {
            if (this._m1obj) {
                this._m1obj.setOutputGain(outputGainMultipler, isDecibel);
            }
        };

        this._m1obj = new(Module).Mach1EncodeInternal();
    };
};
